content/blog/2022-05-20-Banked_Memory_Compiler_Backend_in_MLIR/index.md

```diff
@@ -0,0 +1,302 @@
+"""
+[[extra.authors]]
+name = "Andrew Butt"
+link = "https://andrewb1999.github.io/"  # Links are optional.
+[[extra.authors]]
+name = "Nikita Lazarev"
+link = "https://www.nikita.tech/"  # Links are optional.​
++++
+
+### Code
+
+Unfortunately, as this work is related to ongoing research, we are not yet ready to open-source the project. We have shared the code with Adrian which is available [here](https://www.github.com). If anyone else in the class in interested in seeing the code please reach out and we will give you access as well.
+
+### Introduction
+
+Banked memories, or logical memories that are divided into multiple physical memories, are an important part of achieving high performance FPGA designs. Unfortunately, neither RTL- nor HLS-based languages provide sufficient abstractions to express banked memories. This results in tremendous efforts that programmers need to put when designing memory sub-systems for their hardware architectures. The problem is especially hard given the diverse set of applications with different memory access patterns and performance requirements, as well as numerous hardware primitives the memory can be map onto (e.g. LUTRAM, BRAM, ULTRARAM, etc.) available in today's FPGAs. Without these abstractions, designers have to manually plan the microarchitecture of the memories with all the necessary optimizations in order to make them meeting target design requirements.
+
+To address the aforementioned challenges and simplify the process of designing banked memories, we propose AMC -- a novel MLIR dialect built on top of CIRCT and Calyx that enables compilation of arbitrary memory structures. Our contribution lies in the following three aspects:
+
+* the AMC dialect -- a programming IR abstraction for expressing banked memories;
+* the AMC compiler -- a reference implementation of the AMC dialect built on Calyx;
+* an optimization pass showcasing how our dialect can enable implementation of efficient memory structures when lowering down to hardware.
+
+
+### The AMC dialect
+
+We build the AMC dialect on top of Calyx. Calyx is an IR for compilers generating hardware accelerators. The banked memory specification in the AMC dialect involves four parts: (1) interface specification, (2) memory allocation specification, (3) port specification, and (4) port-interface mapping.
+
+*The memory interface* specifies all read and write interfaces to a unit of banked memory with the corresponding set of arguments. The arguments include the address space size, data width, read/write semantics, and memory access latency in cycles. Example of an interface definition is shown in the listing bellow. Here we define a memory component *@test1* that implements a banked memory with four ports (two on read, two on write), each having 512 32-bit words and 1 cycle of guaranteed access latency.
+```mlir
+amc.memory @test1(!amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>, !amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>);
```

Comment:
While it's all implicit in the prior paragraph I think it might be useful to be explicit about what the three arguments to `amc.port` are: namely, the geometry, direction, and (cycle) latency.